### PR TITLE
Addresses #1955 - Suggests >= y + 1 become > y

### PR DESCRIPTION
**Pull request by mrecachinas**
This is a first pass, and as such, doesn't address a few cases (illustrated below).

This module handles the following cases:
- `... >= ... + 1` and `... >= 1 + ...`
- `... - 1 >= ...` and `-1 + ... >= ...`
- `... + 1 <= ...` and `... + 1 <= ...`
- `... <= ... - 1` and `... <= -1 + ...`

Note: this only goes 1 level deep (i.e., does not constant-fold or 
perform constant propagation) and does not currently simplify 
expressions. 

Examples of these cases include:
```rust
let x = 1;
y >= y + x; // won't catch this case or any permutation

x + 1 >= y + 2; // won't catch this case

x + 1 - 1 >= y - 1 + 1; // WILL catch this case when it likely shouldn't
```

Interested in your thoughts and suggestions. :)

### CODE REVIEW

**Comment by oli-obk** [comment]
Yay! Thanks for starting this lint.

The detection code is perfect. The actual lint call should use `span_lint_and_then` and place a `span_suggestion` with the variable names from the actual code site. For this you'll need to return something more than a Boolean from your detection code though. An `Option<String>` with the replacement text would be one possibility.

**Comment by mrecachinas** [comment]
Sounds good -- I'll make those changes now. Do you think the message should be a bit more specific with respect to adding 1 or subtracting 1, or do you think the general message gets across?

**Comment by oli-obk** [comment]
I think the lint docs are where such details belong. The lint message itself is good

**Comment by mrecachinas** [comment]
@oli-obk Sorry for the delay -- check out the latest commit and let me know your thoughts. I can change it to return `Option<String>` and have the "should I emit a warning" logic handled in the top-most method if you think that would be cleaner.

File: clippy_lints/src/int_plus_one.rs [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+impl IntPlusOne {
+    #[allow(cast_sign_loss)]
+    fn check_lit(&self, lit: &Lit, target_value: i128) -> bool {
+        if let LitKind::Int(value, ..) = lit.node {
+            return value == (target_value as u128)
+        }
+        false
+    }
+
+    fn check_binop(&self, cx: &EarlyContext, block: &Expr, binop: BinOpKind, lhs: &Expr, rhs: &Expr) -> Option<(bool, Option<String>)> {
```
Comment (oli-obk) [inline clippy_lints/src/int_plus_one.rs]: I'm very confused as to how this function successfully compiles ( the return type matches no returned value)

Comment (mrecachinas) [reply clippy_lints/src/int_plus_one.rs]: Good catch -- I started moving toward returning the (bool, String) tuple I described above, but I forgot to remove the return value.

I'm not sure why the compiler didn't catch that though. AFAIK there is no implicit None returned.

**Comment by oli-obk** [comment]
You can remove the bool and inner option and just make the entire return type `Option<String>`

**Comment by mrecachinas** [comment]
If you have any other recommendations to reduce the duplicated code, I'm happy to do it in this PR!

Perhaps having the match statements produce the arguments to `generate_recommendation` and then only call it once?

**Comment by oli-obk** [comment]
I think any further deduplication will only make the code less comprehensible. This looks great now. Let's activate it and see how it goes in the wild :D

**Pull request merged**